File: go/two_pointers/container_with_most_water.go
package two_pointers

func maxArea(heights []int) int {
	n := len(heights)
	left := 0
	right := n - 1
	max := 0

	for left < right {
		if heights[left] < heights[right] {
			min := heights[left]
			value := (right - left) * min
			if value > max {
				max = value
			}
			left++
		} else {
			min := heights[right]
			value := (right - left) * min
			if value > max {
				max = value
			}
			right--
		}
	}

	return max
}

func MaxArea(heights []int) int {
	return maxArea(heights)
}

File: two_pointers/container_with_most_water.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
using namespace std;

class Solution
{
public:
    int maxArea(vector<int> &heights)
    {
        int n = heights.size();
        int max = 0;
        int left = 0;
        int right = n - 1;

        while (left < right)
        {
            int value = (right - left) * this->min(heights[left], heights[right]);
            if (value > max)
            {
                max = value;
            }
            if (this->min(heights[left], heights[right]) == heights[left])
            {
                left++;
            }
            else
            {
                right--;
            }
        }

        return max;
    }
    int min(int x, int y)
    {
        if (x <= y)
        {
            return x;
        }
        else
        {
            return y;
        }
    }
};

int main()
{
    vector<int> inputs = {1, 7, 2, 5, 4, 7, 3, 6};
    Solution *s = new Solution();
    int output = s->maxArea(inputs);
    cout << output << endl;

    return 0;
}